barretenberg/cpp/src/barretenberg/numeric/random/engine.cpp

```diff
@@ -1,17 +1,79 @@
 #include "engine.hpp"
 #include "barretenberg/common/assert.hpp"
 #include <array>
+#include <cstring>
 #include <functional>
+#include <mutex>
 #include <random>
+#include <sys/random.h>
 
 namespace bb::numeric {
 
 namespace {
-auto generate_random_data()
+
+#ifndef __wasm__
+// When working on native we allocate 1M of memory to sample randomness from urandom
+constexpr size_t RANDOM_BUFFER_SIZE = 1UL << 20;
+#else
+// In wasm the API we are using can only give 256 bytes per call, so there is no point in creating a larger buffer
+constexpr size_t RANDOM_BUFFER_SIZE = 256;
+constexpr size_t BYTES_PER_GETENTROPY_READ = 256;
+#endif
+// Buffer with randomness sampled from a CSPRNG
+uint8_t random_buffer[RANDOM_BUFFER_SIZE];
+// Offset into the unused part of the buffer
+ssize_t random_buffer_offset = -1;
+
+#ifndef NO_MULTITHREADING
+// We don't want races to happen
+std::mutex random_buffer_mutex;
+#endif
+
+/**
+ * @brief Generate an array of random unsigned ints sampled from a CSPRNG
+ *
+ * @tparam size_in_unsigned_ints Size of the array
+ * @return std::array<unsigned int, size_in_unsigned_ints>
+ */
+template <size_t size_in_unsigned_ints> std::array<unsigned int, size_in_unsigned_ints> generate_random_data()
 {
-    std::array<unsigned int, 32> random_data;
-    std::random_device source;
-    std::generate(std::begin(random_data), std::end(random_data), std::ref(source));
+    static_assert(size_in_unsigned_ints > 0);
+    static_assert(size_in_unsigned_ints <= 32);
+    std::array<unsigned int, size_in_unsigned_ints> random_data;
+    constexpr size_t random_data_buffer_size = sizeof(random_data);
+
+#ifndef NO_MULTITHREADING
+    std::unique_lock<std::mutex> lock(random_buffer_mutex);
+#endif
+
```

Comment:
This seems like it'd overall work better if the random buffer and random buffer offset were thread_local, then we wouldn't need any locking?

Reply:
Yes, I'll switch. Hope we don't get a situation where we sample very few elements in new threads all the time

Reply:
with the thread pool, that wouldn't be the case